src/Http/WebUtilities/src/MultipartPipeReader.cs

```diff
@@ -0,0 +1,317 @@
+        {
+            _pipeReader = pipeReader ?? throw new ArgumentNullException(nameof(pipeReader));
+            _boundary = new MultipartBoundary(boundary ?? throw new ArgumentNullException(nameof(boundary)), false);
+
+            // This stream will drain any preamble data and remove the first boundary marker. 
+            // TODO: HeadersLengthLimit can't be modified until after the constructor. 
+            _currentSection = new MultipartSectionPipeReader(_pipeReader, _boundary, trackBaseOffsets);
+            _trackBaseOffsets = trackBaseOffsets;
+        }
+
+        /// <summary>
+        /// The limit for the number of headers to read.
+        /// </summary>
+        public int HeadersCountLimit { get; set; } = DefaultHeadersCountLimit;
+
+        /// <summary>
+        /// The combined size limit for headers per multipart section.
+        /// </summary>
+        public int HeadersLengthLimit { get; set; } = DefaultHeadersLengthLimit;
+
+        /// <summary>
+        /// The optional limit for the total response body length.
+        /// </summary>
+        public long? BodyLengthLimit { get; set; }
+
+        public async Task<MultipartSection> ReadNextSectionAsync(CancellationToken cancellationToken = default)
+        {
+            await _currentSection.DrainAsync(cancellationToken);
+            _bytesConsumed += _currentSection.RawLength;
+
+
+            var headersAccumulator = new KeyValueAccumulator();
+            _boundary.ExpectLeadingCrlf = true;
+            long headersLength = 0;
+            while (true)
+            {
+                var readResult = await _pipeReader.ReadAsync(cancellationToken);
+                var buffer = readResult.Buffer;
+
+                if (!buffer.IsEmpty)
+                {
+                    var finishedParsing = TryParseHeadersToEnd(ref buffer, ref headersAccumulator, ref headersLength);
+                    if (headersLength > DefaultHeadersLengthLimit)
+                    {
+                        throw new InvalidDataException($"Multipart headers length limit {HeadersLengthLimit} exceeded.");
+                    }
+                    if (finishedParsing)
+                    {
+                        _bytesConsumed += headersLength;
+                        _pipeReader.AdvanceTo(buffer.Start);
+                        _currentSection = new MultipartSectionPipeReader(_pipeReader, _boundary, _trackBaseOffsets);
+                        long? baseStreamOffset = _trackBaseOffsets ? (long?)_bytesConsumed : null;
+                        return new MultipartSection() { Headers = headersAccumulator.GetResults(), BodyReader = _currentSection, BaseStreamOffset = baseStreamOffset }; ;
+                    }
+                    if (readResult.IsCompleted)
+                    {
+                        throw new InvalidDataException("Unexpected end of Stream, the content may have already been read by another component. ");
```

Comment:
I think AdvanceTo needs to be called before throwing or you'll leave the pipe in an invalid state.

Reply:
will do

Reply:
Wait, why should I call advanceTo on line 91 and not 78 though? or should I call advanceTo in both cases?

Reply:
The lines are now 104 and 91 respectively